---
title: Peak Quality Control
---

::: {.callout-tip}
#### Learning Objectives

- Bulleted list of learning objectives
:::


## Quality Metrics

There are several quality metrics that can be used to assess the quality of the ChIP-seq experiment: 

- These are all output by `nf-core/chipseq`:
  - [Complexity curve](https://www.nature.com/articles/nmeth.2375)
  - [Fingerprint plot](https://deeptools.readthedocs.io/en/latest/content/tools/plotFingerprint.html#background)
  - [Associated fingerprint metrics](https://deeptools.readthedocs.io/en/latest/content/feature/plotFingerprint_QC_metrics.html)
  - [MACS FRiP score](https://www.encodeproject.org/data-standards/terms/)
  - [phantompeakqualtools](https://github.com/kundajelab/phantompeakqualtools) - strand-shift correlation, NSC coefficient, RSC coefficient
  - Correlation between replicates and PCA - in the pipeline this is done with DESeq2 (but `deeptools` can do this also)
- Library complexity (https://www.encodeproject.org/data-standards/terms/#library)
  - PCR Bottlenecking Coefficient 1 (PBC1)
  - PCR Bottlenecking Coefficient 2 (PBC2)
  - Non-Redundant Fraction (NRF)


## Consensus Calling

The way the `nf-core/chipseq` pipeline seems to do the consensus calling is the following: 

- Identify peaks in at least `--min_reps_consensus` replicates per group (a group is a combination of antibody + control, I think)
- Then take the peaks across all the replicates for that antibody and take the _union_ of peaks

An alternative to do consensus calling more "manually" is to call peaks only if they are covered by at least X replicates, following [this tutorial](https://ro-che.info/articles/2018-07-11-chip-seq-consensus).
In this case, the simplest way to do it is to find peaks across all the replicates of an antibody. 
This results in  more peaks than `nf-core/chipseq` gives, because it calls peaks if there was 1 rep in vehicle and 1 rep in e2 with a peak. 
Whereas `nf-core/chipseq` doesn't seem to consider that one as a peak (I checked this by loading the BED consensus from the two approaches on IGV). 

In conclusion: the consensus from `nf-core/chipseq` are probably fine. 


- Annotate peaks:
  - HOMER (again)
- Motif discovery:
  - HOMER


## IGV

- Choose right genome (Human hg38)
- Increase track height, colour, group scale


## Exercise

:::{.callout-exercise}
# Run `nf-core/chipseq`

- Fix the `samplesheet.csv` file where the word "FIXME" appears.
- Fix the script `scripts/01-chipseq_workflow.sh` where the word "FIXME" appears. Output the results to a directory called `results/nf-chipseq`.
- Run the script to execute the workflow. Hint: to run a shell script you use the program `bash name_of_script.sh`
  - Check that the workflow starts running successfully. You should get some progress of the different steps printed on the terminal.
  - The workflow will take some time to run (~20 minutes). You don't need to wait for it to finish before moving on to the next exercise.

:::{.callout-answer collapse=true}


:::

:::


:::{.callout-exercise}
# QC

We've already processed the full dataset through the `nf-core/chipseq` workflow. 
The full data includes 14 samples: 

- 6 samples pulled down with BRD4 antibody: 3 control (vehicle, 'veh') and 3 estrogen-treated ('e2');
- 6 samples pulled down with H2Bub1 antibody: 3 control (vehicle, 'veh') and 3 estrogen-treated ('e2');
- 2 input controls for 'veh' and 'e2' conditions.

Use the file browser to open the MultiQC report generated by the pipeline, which you can find in `preprocessed/nf-chipseq/multiqc/broadPeak/multiqc_report.html`.
Answer the following questions:

- Was the quality of the raw FASTQ files acceptable? Was there any evidence for Illumina adapter contamination? Was this solved after trimming?
- According to the Encode guidelines do you think there were enough mapped reads for our analyses?
- Was there a high percentage of duplicate reads?
- Looking at the fingerprint plot, what can you tell about the differences between the two antibodies? Do you expect different types of peaks?
- From the read distribution profiles, do these two proteins bind to similar regions of annotated genes?
  - Can also look at the annotation from HOMER.
- Something about FRiP score
- Something about peak count
- From PCA do you think treatment had a significant effect on the binding profiles?

Check [ENCODE standards](https://www.encodeproject.org/chip-seq/histone-encode4/#standards).

:::{.callout-answer collapse=true}

For the PCA it's quite interesting to see a correlation between FRiP score and PC1 for BRD4.

:::
:::


## Summary

::: {.callout-tip}
#### Key Points

- Last section of the page is a bulleted summary of the key points
:::
