<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>5&nbsp; Profile heatmaps</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/04-diffbind.html" rel="next">
<link href="../materials/02-peak_ranges.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Analysis of ChIP-seq Data</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="https://docs.google.com/presentation/d/1IuEuwLe7TFbn1kYQmHtsRXvFleSVq2wpaYsymMOEatA/edit?usp=sharing">
 <span class="menu-text">Slides</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/cambiotraining/chipseq"><i class="bi bi-github" role="img" aria-label="Bioinformatics Training Facility GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/bioinfocambs"><i class="bi bi-twitter" role="img" aria-label="Bioinformatics Training Facility Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">
<span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Profile heatmaps</span>
</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Welcome</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data &amp; Setup</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">ChIP Analysis</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/01-peak_calling.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Peak Calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/02-peak_ranges.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploring peak ranges</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/03-profile_heatmaps.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Profile heatmaps</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/04-diffbind.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Differential binding</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#profile-heatmaps" id="toc-profile-heatmaps" class="nav-link active" data-scroll-target="#profile-heatmaps"><span class="toc-section-number">5.1</span>  Profile heatmaps</a>
  <ul class="collapse">
<li><a href="#computematrix" id="toc-computematrix" class="nav-link" data-scroll-target="#computematrix"><span class="toc-section-number">5.1.1</span>  <code>computeMatrix</code></a></li>
  </ul>
</li>
  <li><a href="#profile-plots" id="toc-profile-plots" class="nav-link" data-scroll-target="#profile-plots"><span class="toc-section-number">5.2</span>  Profile plots</a></li>
  <li><a href="#profile-heatmaps-in-r" id="toc-profile-heatmaps-in-r" class="nav-link" data-scroll-target="#profile-heatmaps-in-r"><span class="toc-section-number">5.3</span>  Profile heatmaps in R</a></li>
  <li><a href="#exercise-heatmap" id="toc-exercise-heatmap" class="nav-link" data-scroll-target="#exercise-heatmap"><span class="toc-section-number">5.4</span>  Exercise: heatmap</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title d-none d-lg-block">
<span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Profile heatmaps</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">

</div>
<div class="cell" data-hash="../_cache/purl_a9198e3c3110bfed63b78644b6b7308d">

</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Generate binned count matrices for producing profile plots.</li>
<li>Visualise peak profiles as a density plots and as a heatmap.</li>
</ul>
</div>
</div>
<section id="profile-heatmaps" class="level2" data-number="5.1"><h2 data-number="5.1" class="anchored" data-anchor-id="profile-heatmaps">
<span class="header-section-number">5.1</span> Profile heatmaps</h2>
<p>Before we visualise our data as heatmaps, we need to generate a matrix of binned counts for every peak and around annotations of interest (usually genes). This is a computationally-intense step, so for “real” analysis be sure to run this on a high-performance workstation or cluster (and be ready to wait for a while).</p>
<p>The binned counts matrix can be calculated using the <code>deeptools</code> suite. There are a couple of steps involved:</p>
<ul>
<li>Scale the counts of each sample by the respective input controls. These are usually expressed as a ratio of counts on a log scale: log2(sample/input).</li>
<li>Generate a matrix of binned (scaled) counts around known genes.</li>
</ul>
<p>The first step is achieved using the <code>bamCompare</code> (part of <code>deeptools</code>). Here is how we would generate these counts for one of our samples:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bamCompare</span> <span class="dt">\</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-b1</span> results/nf-chipseq/bwa/mergedLibrary/brd4_e2_rep1.mLb.clN.sorted.bam <span class="dt">\</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-b2</span> results/nf-chipseq/bwa/mergedLibrary/mcf7_input_e2.mLb.clN.sorted.bam <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-o</span> results/deeptools/brd4_e2_rep1.log2.bigwig <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--binSize</span> 20 <span class="dt">\</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--smoothLength</span> 60 <span class="dt">\</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">-p</span> 8</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Here, we have:</p>
<ul>
<li>
<code>-b1</code> is the BAM file for our sample;</li>
<li>
<code>-b2</code> is the BAM file for our input control;</li>
<li>
<code>-o</code> is the output file (in bigwig format);</li>
<li>
<code>--binSize</code> is the size of the bin (in bp) that we want to summarise the data by (default is 10). The larger this number, the “coarser” the output will be, but it will also run faster and produce smaller files.</li>
<li>
<code>--smoothLength</code> will smooth the signal by taking into account the signal of neighbouring regions. This is useful for visualisation.</li>
<li>
<code>-p</code> is the number of processors (CPUs) we want to use to run the analysis in parallel. As mentioned earlier, these are a computationally intensive steps, so using more processors can help speed things up.</li>
</ul>
<p>This analysis would have to be done for every single sample, so you may want to use some programmatic tricks to automate things (see tip box below).</p>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Looping samples through <code>bamCompare</code>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>One way to run <code>bamCompare</code> across all our samples in an automated way is to use a “<a href="https://cambiotraining.github.io/unix-shell/materials/02-programming/03-loops.html">for loop</a>” Below is some code to illustrate this.</p>
<p>We take advantage that our files are named consistently to generate our <code>for</code> loop. In the following example we loop through the E2-treated samples (you would have to write another similar loop for the vehicle samples, as they have a different input control).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> results/nf-chipseq/bwa/mergedLibrary/<span class="pp">*</span>_e2_<span class="pp">*</span>.bam</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get the prefix from the BAM filename to generate output file</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="va">prefix</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .mLb.clN.sorted.bam<span class="va">)</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># print a message to the screen to keep track of things</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="st">"bamCompare on </span><span class="va">${file}</span><span class="st">"</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># run bamCompare</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">bamCompare</span> <span class="dt">\</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">-b1</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">-b2</span> results/nf-chipseq/bwa/mergedLibrary/mcf7_input_e2.mLb.clN.sorted.bam <span class="dt">\</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/deeptools/<span class="va">${prefix}</span>.log2.bigwig <span class="dt">\</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">--binSize</span> 20 <span class="dt">\</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">--smoothLength</span> 60 <span class="dt">\</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">-p</span> 8</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="computematrix" class="level3" data-number="5.1.1"><h3 data-number="5.1.1" class="anchored" data-anchor-id="computematrix">
<span class="header-section-number">5.1.1</span> <code>computeMatrix</code>
</h3>
<p>The next step of our analysis is to create a matrix of binned counts for each peak and gene annotation. This is done using the <code>computeMatrix</code> tool from <code>deeptools</code>.</p>
<p>There are two “modes” in which this tool can compute these counts: <code>reference-point</code> and <code>scale-regions</code>. The difference between these is well illustrated in the <a href="https://deeptools.readthedocs.io/en/latest/content/tools/computeMatrix.html">documentation</a>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="https://deeptools.readthedocs.io/en/latest/_images/computeMatrix_modes.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Difference between the two modes for <code>computeMatrix</code>. Image source: <a href="https://deeptools.readthedocs.io/en/latest/content/tools/computeMatrix.html">deeptools documentation</a>.</figcaption><p></p>
</figure>
</div>
<p>When to use one or the other depends on the type of protein being assessed. For transcription factors, which typically bind to gene promoters, the <code>reference-point</code> mode probably makes sense, as it will show the signal around the TSS. For proteins with broader signals, or that cover gene bodies (like our H2Bub1), the <code>scale-regions</code> might make more sense instead.</p>
<p>Here is some the code we used to generate this matrix, using as input the scaled bigwig files from <code>bamCompare</code>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">computeMatrix</span> scale-regions <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--regionsFileName</span> resources/GRCh38.109.gtf.gz <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--scoreFileName</span> results/deeptools/brd4_<span class="pp">*</span>.log2.bigwig <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFileName</span> results/deeptools/brd4.log2.mat.gz <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--blackListFileName</span> resources/ENCFF356LFX_exclusion_lists.bed.gz <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--upstream</span> 2000 <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--downstream</span> 2000 <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--skipZeros</span> <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--binSize</span> 20 <span class="dt">\</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">-p</span> 8</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Where:</p>
<ul>
<li>
<code>--regionsFileName</code> is the name of the annotations we want to use to estimate the matrix of counts. Typically the GTF of gene annotations is used (but you could use a different annotation, e.g.&nbsp;for a small subset of genes of interest).</li>
<li>
<code>--scoreFileName</code> is the name of the files we want to calculate the binned counts on. Notice how we use the <code>*</code> wildcard to match all the BRD4.</li>
<li>
<code>--outFileName</code> is the name for the output file, which is automatically compressed using the ‘gzip’ algorithm (thus we save it with extension <code>.gz</code>).</li>
<li>
<code>--blackListFileName</code> excludes regions falling in our exclusions list from being processed.</li>
<li>
<code>--skipZeros</code> skips regions that have no counts.</li>
<li>
<code>--upstream</code> and <code>--downstream</code> is the number of bp upstream and downstream of the TSS that we want to summarise counts, respectively.</li>
<li>
<code>--binSize</code> and <code>-p</code> have the same meaning as above for <code>bamCompare</code>.</li>
</ul>
<p>This step takes quite a while to run, so we have already generated these files for you (available from <code>preprocessed/deeptools/</code>).</p>
</section></section><section id="profile-plots" class="level2" data-number="5.2"><h2 data-number="5.2" class="anchored" data-anchor-id="profile-plots">
<span class="header-section-number">5.2</span> Profile plots</h2>
<p>There are two main types of profile plots that we can do:</p>
<ul>
<li>Density plots showing the average signal of our peaks around our regions of interest (often genes).</li>
<li>A heatmap showing the signal of individual peaks around those same regions.</li>
</ul>
<p>Once we have our binned count matrix, generating these is relatively simple using two <code>deeptools</code> commands.</p>
<p>For profile density plots we can use:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plotProfile</span> <span class="at">-m</span> preprocessed/deeptools/brd4.chr21.mat.gz <span class="at">-out</span> results/brd4_profile.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And for heatmaps we can do:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plotHeatmap</span> <span class="at">-m</span> preprocessed/deeptools/brd4.chr21.mat.gz <span class="at">-out</span> results/brd4_heatmap.png</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>There are several other options available with both of these tools (look at their documentation: <a href="https://deeptools.readthedocs.io/en/latest/content/tools/plotProfile.html">here</a> and <a href="https://deeptools.readthedocs.io/en/latest/content/tools/plotHeatmap.html">here</a>), including the ability to cluster the peaks using K-means clustering. We will leave that as an optional exercise for the reader.</p>
</section><section id="profile-heatmaps-in-r" class="level2" data-number="5.3"><h2 data-number="5.3" class="anchored" data-anchor-id="profile-heatmaps-in-r">
<span class="header-section-number">5.3</span> Profile heatmaps in R</h2>
<p>Although deeptools makes generating these plots relatively simple, we can also produce similar analysis using dedicated R/Bioconductor packages. This has the advantage that it integrates well with our previous analysis in R.</p>
<p>First, we set things up by loading our packages:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Packages ----</span></span>
<span></span>
<span><span class="co"># load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span> <span class="co"># for importing BED/GFF/etc.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span>   <span class="co"># for working with GenomicRanges</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ChIPseeker</span><span class="op">)</span>  <span class="co"># to annotate peaks</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">profileplyr</span><span class="op">)</span> <span class="co"># for profile heatmaps</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># change the default ggplot theme</span></span>
<span><span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_classic</span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you haven’t run the previous analysis of peaks, you can also read these in from pre-processed files we provide:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># read consensus GRanges</span></span>
<span><span class="va">brd4_consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"preprocessed/r_objects/brd4_consensus_granges.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The matrix of binned counts can be imported into R using the <code>Bioconductor/profileplyr</code> package.</p>
<div class="cell" data-hash="../_cache/import_profile_c1cff1eb9eff82752d09cc56de7d64f1">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Profile heatmaps ----</span></span>
<span></span>
<span><span class="co"># import the pre-computed matrix from deeptools</span></span>
<span><span class="va">brd4_prof</span> <span class="op">&lt;-</span> <span class="fu">import_deepToolsMat</span><span class="op">(</span><span class="st">"preprocessed/deeptools/brd4.log2.chr21.mat.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">brd4_prof</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: profileplyr 
dim: 3138 250 
metadata(0):
assays(6): brd4_e2_rep1.log2 brd4_e2_rep2.log2 ... brd4_veh_rep2.log2
  brd4_veh_rep3.log2
rownames: NULL
rowData names(3): names score dpGroup
colnames: NULL
colData names(0):</code></pre>
</div>
</div>
<p>The type of object returned is called a <code>profileplyr</code> object, which is a modified version of a <code>SummarizedExperiment</code> object, often used to store RNA-seq counts. We won’t go into the details of this object here, but you can find detailed information on the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/profileplyr/inst/doc/profileplyr.html#the-profileplyr-object">package’s vignette</a>.</p>
<p>The main things to consider for our purposes are:</p>
<ul>
<li>The peak intervals (ranges) over which the data are summarised is available through the <code>rowRanges()</code> slot of the object.</li>
<li>The sample information is stored in the <code>sampleData()</code> slot of the object.</li>
<li>We can subset the object using <code>[</code> with the syntax as: <code>object[ranges, bins, samples]</code>.</li>
</ul>
<p>Here are some illustrative examples:</p>
<div class="cell" data-hash="../_cache/subset_profiler_ae366b9d6e61698acabdd588d42ee158">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ranges information</span></span>
<span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">brd4_prof</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3138 ranges and 3 metadata columns:
         seqnames            ranges strand |           names       score
            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;
     [1]       21   8238858-8239092      + | ENST00000652328           .
     [2]       21   8420674-8421003      + | ENST00000652246           .
     [3]       21 10576291-10576587      + | ENST00000616920           .
     [4]       21 10028360-10029855      - | ENST00000624662           .
     [5]       21 13544209-13545185      - | ENST00000431829           .
     ...      ...               ...    ... .             ...         ...
  [3134]       21 25934340-25935178      + | ENST00000596669           .
  [3135]       21 25941601-25952614      + | ENST00000667606           .
  [3136]       21 25943043-25943145      - | ENST00000516163           .
  [3137]       21 26158090-26175824      + | ENST00000455275           .
  [3138]       21 26190706-26190813      + | ENST00000384075           .
          dpGroup
         &lt;factor&gt;
     [1]    genes
     [2]    genes
     [3]    genes
     [4]    genes
     [5]    genes
     ...      ...
  [3134]    genes
  [3135]    genes
  [3136]    genes
  [3137]    genes
  [3138]    genes
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># sample information</span></span>
<span><span class="fu">sampleData</span><span class="op">(</span><span class="va">brd4_prof</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 20 columns
                    upstream downstream      body  bin.size unscaled.5.prime
                   &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;        &lt;numeric&gt;
brd4_e2_rep1.log2       2000       2000      1000        20                0
brd4_e2_rep2.log2       2000       2000      1000        20                0
brd4_e2_rep3.log2       2000       2000      1000        20                0
brd4_veh_rep1.log2      2000       2000      1000        20                0
brd4_veh_rep2.log2      2000       2000      1000        20                0
brd4_veh_rep3.log2      2000       2000      1000        20                0
                   unscaled.3.prime      sample_labels   verbose bin.avg.type
                          &lt;numeric&gt;        &lt;character&gt; &lt;logical&gt;  &lt;character&gt;
brd4_e2_rep1.log2                 0  brd4_e2_rep1.log2     FALSE         mean
brd4_e2_rep2.log2                 0  brd4_e2_rep2.log2     FALSE         mean
brd4_e2_rep3.log2                 0  brd4_e2_rep3.log2     FALSE         mean
brd4_veh_rep1.log2                0 brd4_veh_rep1.log2     FALSE         mean
brd4_veh_rep2.log2                0 brd4_veh_rep2.log2     FALSE         mean
brd4_veh_rep3.log2                0 brd4_veh_rep3.log2     FALSE         mean
                   missing.data.as.zero     scale skip.zeros nan.after.end
                              &lt;logical&gt; &lt;numeric&gt;  &lt;logical&gt;     &lt;logical&gt;
brd4_e2_rep1.log2                 FALSE         1       TRUE         FALSE
brd4_e2_rep2.log2                 FALSE         1       TRUE         FALSE
brd4_e2_rep3.log2                 FALSE         1       TRUE         FALSE
brd4_veh_rep1.log2                FALSE         1       TRUE         FALSE
brd4_veh_rep2.log2                FALSE         1       TRUE         FALSE
brd4_veh_rep3.log2                FALSE         1       TRUE         FALSE
                   proc.number sort.regions  sort.using ref.point min.threshold
                     &lt;numeric&gt;  &lt;character&gt; &lt;character&gt;    &lt;list&gt;        &lt;list&gt;
brd4_e2_rep1.log2           10         keep        mean                        
brd4_e2_rep2.log2           10         keep        mean                        
brd4_e2_rep3.log2           10         keep        mean                        
brd4_veh_rep1.log2          10         keep        mean                        
brd4_veh_rep2.log2          10         keep        mean                        
brd4_veh_rep3.log2          10         keep        mean                        
                   max.threshold generation.method
                          &lt;list&gt;       &lt;character&gt;
brd4_e2_rep1.log2                        deepTools
brd4_e2_rep2.log2                        deepTools
brd4_e2_rep3.log2                        deepTools
brd4_veh_rep1.log2                       deepTools
brd4_veh_rep2.log2                       deepTools
brd4_veh_rep3.log2                       deepTools</code></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># subset the object for:</span></span>
<span><span class="co"># first 10 peaks</span></span>
<span><span class="co"># first 5 bins</span></span>
<span><span class="co"># first 3 samples</span></span>
<span><span class="va">brd4_prof</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: profileplyr 
dim: 10 5 
metadata(0):
assays(3): brd4_e2_rep1.log2 brd4_e2_rep2.log2 brd4_e2_rep3.log2
rownames: NULL
rowData names(3): names score dpGroup
colnames: NULL
colData names(0):</code></pre>
</div>
</div>
<p>One of the main functions provided in this package is the ability to produce profile heatmaps. These heatmaps can be quite heavy to render graphically, so it’s usually a good idea to subset our ranges.</p>
<p>In the following example, we sample a few random intervals to get an idea of occupancy for our BRD4 protein around genes:</p>
<div class="cell" data-hash="../_cache/enriched_heatmap_fce6e40f9e2a262c70e519ebecd80c32">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># sample 200 peaks randomly</span></span>
<span><span class="va">random_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">brd4_prof</span><span class="op">)</span>, <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot random peaks to give an idea of occupancy</span></span>
<span><span class="fu">generateEnrichedHeatmap</span><span class="op">(</span><span class="va">brd4_prof</span><span class="op">[</span><span class="va">random_peaks</span>, <span class="op">]</span>, </span>
<span>                        include_group_annotation <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="03-profile_heatmaps_files/figure-html/enriched_heatmap-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can see that BRD4 mostly occupies gene promoters. This result does not come as a complete surprise, as we had already seen this in previous analysis.</p>
<p>For heatmaps with thousands of peaks, it’s best to save them as PNG, as the plotting device on RStudio can stuggle to render them. Here is an example for all Chr21 peaks we loaded in:</p>
<div class="cell" data-hash="../_cache/unnamed-chunk-1_03197c2d96532dce98450240e955ab05">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># save as PNG - adjust width, height and resolution to fit your needs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"results/brd4_heatmap.png"</span>, width <span class="op">=</span> <span class="fl">3000</span>, height <span class="op">=</span> <span class="fl">1500</span>, res <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="fu">generateEnrichedHeatmap</span><span class="op">(</span><span class="va">brd4_prof</span>, include_group_annotation <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can subset our profile object to include only the peaks that overlap with another set of peaks of our interest. For example, we could investigate if the profiles look very different for the set of genes that are differentially expressed in siBRD4 lines.</p>
<div class="cell" data-hash="../_cache/profile_degs_0273a2c2c825d6758494ba1465b67033">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># read DEGs from Nagarajan 2017</span></span>
<span><span class="va">degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"resources/degs_nagarajan2017.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset annotated intervals </span></span>
<span><span class="va">brd4_consensus_degs</span> <span class="op">&lt;-</span> <span class="va">brd4_consensus</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">geneId</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">degs</span><span class="op">$</span><span class="va">ensembl_gene_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset our profile object to include only the filtered intervals</span></span>
<span><span class="va">brd4_prof_degs</span> <span class="op">&lt;-</span> <span class="va">brd4_prof</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">subsetByOverlaps</span><span class="op">(</span><span class="va">brd4_consensus_degs</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># visualise them</span></span>
<span><span class="fu">generateEnrichedHeatmap</span><span class="op">(</span><span class="va">brd4_prof_degs</span>, include_group_annotation <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="03-profile_heatmaps_files/figure-html/profile_degs-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="exercise-heatmap" class="level2" data-number="5.4"><h2 data-number="5.4" class="anchored" data-anchor-id="exercise-heatmap">
<span class="header-section-number">5.4</span> Exercise: heatmap</h2>
<div class="callout-exercise callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Based on the code we showed above, generate a profile heatmap for the H2Bub1 ChIP.</p>
<ul>
<li>First import the matrix of binned counts using <code>import_deepToolsMat()</code> and save it into an object called <code>h2bub1_prof</code>.</li>
<li>Then generate a heatmap using the <code>generateEnrichedHeatmap()</code> function, making sure you save it as a PNG.</li>
</ul>
<p>Does the result match your expectation from previous analysis?</p>
<div class="cell" data-hash="../_cache/ex1_h2bub1_prof_e0c978e01a0da92261f0c9b71e71b938">

</div>
<div class="callout-answer callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Here is the code to import and generate a PNG file with our heatmap:</p>
<div class="cell" data-hash="../_cache/ex_answer_heatmap_d3f63447e8a5ccc0511e268089f12fda">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># import the pre-computed matrix from deeptools</span></span>
<span><span class="va">h2bub1_prof</span> <span class="op">&lt;-</span> <span class="fu">import_deepToolsMat</span><span class="op">(</span><span class="st">"preprocessed/deeptools/h2bub1.log2.chr21.mat.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save plot as PNG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span><span class="st">"results/h2bub1_heatmap.png"</span>, width <span class="op">=</span> <span class="fl">3000</span>, height <span class="op">=</span> <span class="fl">1500</span>, res <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="fu">generateEnrichedHeatmap</span><span class="op">(</span><span class="va">h2bub1_prof</span>, include_group_annotation <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In order to save space, we show a random subset of 200 genes below as an example:</p>
<div class="cell" data-hash="../_cache/ex_answer_heatmap_part2_cc30e74bbaa34f6d3d4c297b3ed08b17">
<div class="cell-output-display">
<p><img src="03-profile_heatmaps_files/figure-html/ex_answer_heatmap_part2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The plot shows that H2Bub1 has broader peaks across whole gene bodies and promoters. This fits with the the peak annotation we did earlier, which showed that most peaks occur in those regions.</p>
</div>
</div>
</div>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../materials/02-peak_ranges.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploring peak ranges</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../materials/04-diffbind.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Differential binding</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
      <div class="nav-footer-center">Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Bioinformatics Training Facility</div>
  </div>
</footer>


</body></html>