<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>6&nbsp; Differential binding</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../materials/03-profile_heatmaps.html" rel="prev">
<link href="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university-of-cambridge-favicon.ico" rel="icon">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../site_libs/quarto-contrib/quarto-project/cambiotraining/courseformat/img/university_crest_reversed.png" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Analysis of ChIP-seq Data</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="https://docs.google.com/presentation/d/1IuEuwLe7TFbn1kYQmHtsRXvFleSVq2wpaYsymMOEatA/edit?usp=sharing">
 <span class="menu-text">Slides</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/cambiotraining/chipseq"><i class="bi bi-github" role="img" aria-label="Bioinformatics Training Facility GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/bioinfocambs"><i class="bi bi-twitter" role="img" aria-label="Bioinformatics Training Facility Twitter">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }"><div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">
<span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Differential binding</span>
</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Welcome</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../setup.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data &amp; Setup</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">ChIP Analysis</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/01-peak_calling.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Peak Calling</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/02-peak_ranges.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Exploring peak ranges</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/03-profile_heatmaps.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Profile heatmaps</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../materials/04-diffbind.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Differential binding</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><!-- margin-sidebar --><div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#differential-binding" id="toc-differential-binding" class="nav-link active" data-scroll-target="#differential-binding"><span class="toc-section-number">6.1</span>  Differential binding</a></li>
  <li><a href="#reading-peaks" id="toc-reading-peaks" class="nav-link" data-scroll-target="#reading-peaks"><span class="toc-section-number">6.2</span>  Reading peaks</a></li>
  <li><a href="#differential-binding-visualisation" id="toc-differential-binding-visualisation" class="nav-link" data-scroll-target="#differential-binding-visualisation"><span class="toc-section-number">6.3</span>  Differential binding visualisation</a></li>
  <li><a href="#pipeline" id="toc-pipeline" class="nav-link" data-scroll-target="#pipeline"><span class="toc-section-number">6.4</span>  Pipeline</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="toc-section-number">6.5</span>  Exercises</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title d-none d-lg-block">
<span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Differential binding</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell">

</div>
<div class="cell" data-hash="../_cache/unnamed-chunk-1_2c879d4d624a0e1d51d3d372b1bffdbd">

</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Apply the <code>DiffBind</code> package to identify differentially bound peaks between two conditions.</li>
</ul>
</div>
</div>
<section id="differential-binding" class="level2" data-number="6.1"><h2 data-number="6.1" class="anchored" data-anchor-id="differential-binding">
<span class="header-section-number">6.1</span> Differential binding</h2>
<p>When multiple conditions are available in our ChIP design (e.g.&nbsp;healthy vs disease, wild-type vs mutant, treated vs control, different cell types, …), we may want to investigate whether certain peaks are differentially bound between them.</p>
<p>Finding differentially bound regions in the genome is analogous to identifying differentially expressed genes in RNA-seq data. In both cases we are dealing with count data summarised over features (genes/transcripts in the case of RNA-seq and peaks in the case of ChIP-seq). In both cases the biological replicates show larger variability than technical replicates, and the negative binomial model is suitable to compare binding affinities accross samples.</p>
<p>There are a number of differential expression packages in R that use the negative binomial model e.g.&nbsp;<code>DESeq2</code> and <code>edgeR</code>. These methods are wrapped in the <code>DiffBind</code> package that is geared towards analysing differential binding in ChIP-seq data and provides a number of analytical plots as well.</p>
<p>As usual, we start by loading our packages:</p>
<div class="cell" data-hash="../_cache/packages_1c1ba9bf62d0364cf2949ab3b12975c5">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Packages ----</span></span>
<span></span>
<span><span class="co"># load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span> <span class="co"># for importing BED/GFF/etc.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span>   <span class="co"># for working with GenomicRanges</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ChIPseeker</span><span class="op">)</span>  <span class="co"># to annotate peaks</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">profileplyr</span><span class="op">)</span> <span class="co"># for profile heatmaps</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">DiffBind</span><span class="op">)</span>    <span class="co"># for ChIP peak analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># change the default ggplot theme</span></span>
<span><span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_classic</span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="reading-peaks" class="level2" data-number="6.2"><h2 data-number="6.2" class="anchored" data-anchor-id="reading-peaks">
<span class="header-section-number">6.2</span> Reading peaks</h2>
<p>As is common with Bioconductor packages, the <code>DiffBind</code> package uses its own data structure, called a <code>dba</code> object. To create this, we need to create a CSV file with the following information:</p>
<ul>
<li>TODO long list of columns</li>
</ul>
<p>We’ve already created this for you, so all we need to do is read it in:</p>
<div class="cell" data-hash="../_cache/read_samplesheet_95a1b971a85c0a4d0a8cf4867c5708a9">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># read DiffBind samplesheet</span></span>
<span><span class="va">samplesheet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"diffbind_samplesheet.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">samplesheet</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SampleID                   Tissue Antibody Condition Treatment Replicate
1  brd4_e2_rep1 MCF7 breast cancer cells     BRD4        NA        e2         1
2  brd4_e2_rep2 MCF7 breast cancer cells     BRD4        NA        e2         2
3  brd4_e2_rep3 MCF7 breast cancer cells     BRD4        NA        e2         3
4 brd4_veh_rep1 MCF7 breast cancer cells     BRD4        NA       veh         1
5 brd4_veh_rep2 MCF7 breast cancer cells     BRD4        NA       veh         2
6 brd4_veh_rep3 MCF7 breast cancer cells     BRD4        NA       veh         3
                                                                    bamReads
1  preprocessed/nf-chipseq/bwa/mergedLibrary/brd4_e2_rep1.mLb.clN.sorted.bam
2  preprocessed/nf-chipseq/bwa/mergedLibrary/brd4_e2_rep2.mLb.clN.sorted.bam
3  preprocessed/nf-chipseq/bwa/mergedLibrary/brd4_e2_rep3.mLb.clN.sorted.bam
4 preprocessed/nf-chipseq/bwa/mergedLibrary/brd4_veh_rep1.mLb.clN.sorted.bam
5 preprocessed/nf-chipseq/bwa/mergedLibrary/brd4_veh_rep2.mLb.clN.sorted.bam
6 preprocessed/nf-chipseq/bwa/mergedLibrary/brd4_veh_rep3.mLb.clN.sorted.bam
       ControlID
1  mcf7_input_e2
2  mcf7_input_e2
3  mcf7_input_e2
4 mcf7_input_veh
5 mcf7_input_veh
6 mcf7_input_veh
                                                                   bamControl
1  preprocessed/nf-chipseq/bwa/mergedLibrary/mcf7_input_e2.mLb.clN.sorted.bam
2  preprocessed/nf-chipseq/bwa/mergedLibrary/mcf7_input_e2.mLb.clN.sorted.bam
3  preprocessed/nf-chipseq/bwa/mergedLibrary/mcf7_input_e2.mLb.clN.sorted.bam
4 preprocessed/nf-chipseq/bwa/mergedLibrary/mcf7_input_veh.mLb.clN.sorted.bam
5 preprocessed/nf-chipseq/bwa/mergedLibrary/mcf7_input_veh.mLb.clN.sorted.bam
6 preprocessed/nf-chipseq/bwa/mergedLibrary/mcf7_input_veh.mLb.clN.sorted.bam
                                                                                    Peaks
1  preprocessed/nf-chipseq/bwa/mergedLibrary/macs2/broadPeak/brd4_e2_rep1_peaks.broadPeak
2  preprocessed/nf-chipseq/bwa/mergedLibrary/macs2/broadPeak/brd4_e2_rep2_peaks.broadPeak
3  preprocessed/nf-chipseq/bwa/mergedLibrary/macs2/broadPeak/brd4_e2_rep3_peaks.broadPeak
4 preprocessed/nf-chipseq/bwa/mergedLibrary/macs2/broadPeak/brd4_veh_rep1_peaks.broadPeak
5 preprocessed/nf-chipseq/bwa/mergedLibrary/macs2/broadPeak/brd4_veh_rep2_peaks.broadPeak
6 preprocessed/nf-chipseq/bwa/mergedLibrary/macs2/broadPeak/brd4_veh_rep3_peaks.broadPeak
  PeakCaller
1        bed
2        bed
3        bed
4        bed
5        bed
6        bed</code></pre>
</div>
</div>
<p>Once we have this samplesheet, creating the <code>dba</code> object is relatively simple (note in this case we’re only loading the BRD4 samples):</p>
<div class="cell" data-hash="../_cache/read_dba_a679bdf5cfa9fc5543e074f4707b6458">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create DBA object for BRD4 antibody</span></span>
<span><span class="va">brd4_dba</span> <span class="op">&lt;-</span> <span class="fu">dba</span><span class="op">(</span>sampleSheet <span class="op">=</span> <span class="va">samplesheet</span><span class="op">[</span><span class="va">samplesheet</span><span class="op">$</span><span class="va">Antibody</span> <span class="op">==</span> <span class="st">"BRD4"</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">brd4_dba</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>6 Samples, 12395 sites in matrix (48128 total):
             ID                   Tissue Treatment Replicate Intervals
1  brd4_e2_rep1 MCF7 breast cancer cells        e2         1     20240
2  brd4_e2_rep2 MCF7 breast cancer cells        e2         2     11144
3  brd4_e2_rep3 MCF7 breast cancer cells        e2         3      2599
4 brd4_veh_rep1 MCF7 breast cancer cells       veh         1      2922
5 brd4_veh_rep2 MCF7 breast cancer cells       veh         2     37306
6 brd4_veh_rep3 MCF7 breast cancer cells       veh         3      2107</code></pre>
</div>
</div>
<p>Without doing any further analysis, we can already produce a correlation heatmap using the <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code> function:</p>
<div class="cell" data-hash="../_cache/plot_raw_5d10c6f3d6866565e42e2fa8de0a4cd3">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># correlation plot using caller score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">brd4_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/plot_raw-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This correlation is calculated based on the score given to each peak, in our case the score that MACS assigns to each called peak. This is not necessarily the best way to look at the correlation between samples, as it doesn’t take into account the actual counts in each peak (that will come later), but it’s a good starting point to look at the correlation between our samples.</p>
<p>In our case, we can see that our replicates are not clustering per treatment, suggesting other effects may have played a role in the ChIP profiles.</p>
<p>The next step is to count reads in each peak (which is the raw data that will be used to estimate the differential binding). This step takes a long time, so be prepared to wait if you’re running this on your data.</p>
<div class="cell" data-hash="../_cache/count_436bdf494b5fe82dc76a529b26f72900">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># count reads overlapping peaks </span></span>
<span><span class="co"># this takes a long time to run! So we load pre-computed one</span></span>
<span><span class="co"># brd4_dba &lt;- dba.count(brd4_dba)</span></span>
<span><span class="va">brd4_dba</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"preprocessed/r_objects/brd4_dba.count.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># correlation plot based on raw counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">brd4_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/count-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now, when we call the <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code> function the correlation heatmap is instead done from the counts assigned to each peak (rather than peak scores). In this case, we get a clearer clustering of samples by treatment, which is a good sign!</p>
<p>The next step in the analysis is to normalise the counts. There are different methods to normalise the counts, and this is discussed at length by the DiffBind authors (see section 7 in the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf">documentation</a>). In summary, these authors argue for caution when applying normalisation methods used for RNA-seq analysis, as those methods assume that most features (peaks in our case, genes in the RNA-seq case) are <em>not</em> differentially expressed/bound between conditions. However, this assumption may not make biological sense in the case of ChIP-seq, as some conditions may dramatically affect the binding profiles genome-wide.</p>
<p>This may certainly be the case for BRD4, as we’ve seen from our profile plots that binding is generally higher for E2-treated samples compared to controls. For this reason, the default normalisation method used by <code>DiffBind</code> is a simple library size normalisation, which we apply below.</p>
<div class="cell" data-hash="../_cache/normalise_2d021557b532d8bbe414e1c326997d15">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># normalise counts by library size (default)</span></span>
<span><span class="va">brd4_dba</span> <span class="op">&lt;-</span> <span class="fu">dba.normalize</span><span class="op">(</span><span class="va">brd4_dba</span>, normalize <span class="op">=</span> <span class="va">DBA_NORM_LIB</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One final step before we fit the model that will test for differential binding, is to set the “contrast” of our experiment. This is where we can set which factors should be taken into account when modelling the counts and which comparisons we want to make. There is quite a lot of flexibility in setting these models, which is detailed in the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf">package’s documentation</a> as well as <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">the documentation for <code>DEseq2</code></a>, which is used behind the scenes.</p>
<p>In our case, we only have one factor (treatment) with two conditions (“e2” and “veh”). Therefore, we can set our contrast like so:</p>
<div class="cell" data-hash="../_cache/contrast_5a2d1af2eecc7478b18a1adbd4bb672e">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># set contrast </span></span>
<span><span class="va">brd4_dba</span> <span class="op">&lt;-</span> <span class="fu">dba.contrast</span><span class="op">(</span><span class="va">brd4_dba</span>, </span>
<span>                         design <span class="op">=</span> <span class="op">~</span> <span class="va">Treatment</span>,</span>
<span>                         reorderMeta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="st">"veh"</span><span class="op">)</span>,</span>
<span>                         minMembers <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Where:</p>
<ul>
<li>
<code>design</code> uses R’s standard model formula syntax to define the variables used to model our counts, in our case “Treatment” (this is one of the columns from our CSV samplesheet).</li>
<li>
<code>reorderMeta</code> is used to set the reference level for our Treatment variable; in our case it makes sense to set “veh” as the reference level (control). The default reference level would be whichever comes first alphabetically.</li>
<li>
<code>minMembers</code> defines the minimum number of replicates required to run the analysis. The default value is in fact 3, but we’ve set it anyway to be explicit.</li>
</ul>
<p>Finally, we can fit the statical model using <code>DEseq2</code>’s statistical machinery (the other option is to use <code>DBA_EDGER</code>, which would use that package instead - the results are very comparable).</p>
<div class="cell" data-hash="../_cache/analysis_82477774baa396279e7b9d54da957c58">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># run the analysis</span></span>
<span><span class="va">brd4_dba</span> <span class="op">&lt;-</span> <span class="fu">dba.analyze</span><span class="op">(</span><span class="va">brd4_dba</span>, </span>
<span>                        method <span class="op">=</span> <span class="va">DBA_DESEQ2</span>,</span>
<span>                        bBlacklist <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                        bGreylist <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that we’ve set two options to <code>FALSE</code>:</p>
<ul>
<li>
<code>bBlacklist</code> turns off the behaviour of trying to identify our genome and apply an exclusion list to it. We’ve already done this when we did peak calling with the <code>nf-core/chipseq</code> workflow, so there is no need to do this again.</li>
<li>
<code>bGreylist</code> turns off the behaviour of estimating a so-called “greylist”. We talk more about this below.</li>
</ul>
<p>Finally, with the analysis done, it’s time to extract our results, which are returned as a familiar <code>GRanges</code> object:</p>
<div class="cell" data-hash="../_cache/report_2fe018f2af363f0497dc82612358dea1">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># extract diffbound sites</span></span>
<span><span class="co"># keep all peaks, even those that are non-significant</span></span>
<span><span class="va">brd4_diffbound</span> <span class="op">&lt;-</span> <span class="fu">dba.report</span><span class="op">(</span><span class="va">brd4_dba</span>, th <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">brd4_diffbound</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 12331 ranges and 6 metadata columns:
          seqnames            ranges strand |      Conc   Conc_e2  Conc_veh
             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
   7853         21 42365357-42365757      * |   4.56334   5.53382   0.00000
   7548         20 53865352-53865752      * |   5.10739   2.25210   6.00409
   7493         20 53660691-53661091      * |   4.63297   5.54957   1.47889
   3675         15 74816468-74816868      * |   4.46242   5.40868   0.68917
   7263         20 44714529-44714929      * |   3.12993   4.12993   0.00000
    ...        ...               ...    ... .       ...       ...       ...
  12064 KI270754.1         3352-3752      * |   1.47054   1.44409   1.49652
   7710         20 63330620-63331020      * |   1.41217   1.40120   1.42307
   1682         11 31509716-31510116      * |   1.53077   1.53102   1.53053
   5879         19 16535344-16535744      * |   1.55841   1.55877   1.55805
  12034 KI270467.1         3240-3640      * |   5.08000   4.95839   5.19215
                Fold     p-value        FDR
           &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;
   7853      4.43696 8.71201e-07 0.00756452
   7548     -3.35432 1.33026e-06 0.00756452
   7493      3.47205 5.27113e-06 0.01612934
   3675      3.94028 5.67285e-06 0.01612934
   7263      4.44204 8.38315e-06 0.01906832
    ...          ...         ...        ...
  12064 -0.000219130    0.994758          1
   7710 -0.000213502    0.994952          1
   1682  0.000186362    0.995030          1
   5879  0.000171415    0.995530          1
  12034 -0.002416403    1.000000          1
  -------
  seqinfo: 44 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>We can use some <code>plyranges</code> syntax to obtain a summary of how many peaks are differentially bound and in which direction:</p>
<div class="cell" data-hash="../_cache/count_diffbound_375721f218ac13b660a86bc05bfbfebd">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># count how many up or down</span></span>
<span><span class="va">brd4_diffbound</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">summarise</span><span class="op">(</span>up <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">Fold</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>, down <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">Fold</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 1 row and 2 columns
         up      down
  &lt;integer&gt; &lt;integer&gt;
1        30         4</code></pre>
</div>
</div>
<p>The result is not particularly striking. It suggests that not many peaks are differentially bound between conditions. Let’s investigate our results a bit further with some visualisations.</p>
</section><section id="differential-binding-visualisation" class="level2" data-number="6.3"><h2 data-number="6.3" class="anchored" data-anchor-id="differential-binding-visualisation">
<span class="header-section-number">6.3</span> Differential binding visualisation</h2>
<div class="cell" data-hash="../_cache/header_b2560d672742f9454b863c6a6a07772c">

</div>
<p>The <code>DiffBind</code> package provides several plotting functions, illustrated below.</p>
<p>We start with a PCA, which shows that samples separate by treatment along PC2, but not PC1 (the axis of greatest variance). This suggests that factors other than the treatment influenced our ChIP results. This may explain the relatively low number of differentially bound peaks.</p>
<div class="cell" data-hash="../_cache/plot_pca_4ef43ef3046a354a26eaac6cb2f76b7b">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># PCA plot</span></span>
<span><span class="fu">dba.plotPCA</span><span class="op">(</span><span class="va">brd4_dba</span>, label <span class="op">=</span> <span class="va">DBA_REPLICATE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/plot_pca-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Another common visualisation used in differential analysis is the MA plot. This shows the average normalised read counts on the X-axis and the log-fold change on the Y-axis. This plot can be used to assess the effect of the normalisation on the data (red trend line, which we want to be close to the zero horizontal line) as well as highlighting the significant differentially bound regions (pink points).</p>
<div class="cell" data-hash="../_cache/plot_ma_3ea5461abe54633c83a4149a9539f4bb">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># MA plot</span></span>
<span><span class="fu">dba.plotMA</span><span class="op">(</span><span class="va">brd4_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/plot_ma-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>You can also generate this plot with <code>ggplot2</code>, if you want to customise it further:</p>
<div class="cell" data-hash="../_cache/unnamed-chunk-2_1ab8ba03825f45badfc3e2133b016ace">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># can also do it with ggplot2</span></span>
<span><span class="va">brd4_diffbound</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>sig <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">Fold</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">Conc</span>, <span class="va">Fold</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">sig</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Another common visualisation is a volcano plot, which shows the log-fold change on the x-axis and p-values on the y-axis (we don’t show this one, to save space).</p>
<div class="cell" data-hash="../_cache/plot_volcano_41e6fac8b8d966edaf8b9783d7a685db">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># volcano plot</span></span>
<span><span class="fu">dba.plotVolcano</span><span class="op">(</span><span class="va">brd4_dba</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># can also do it with ggplot2</span></span>
<span><span class="va">brd4_diffbound</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>sig <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">Fold</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">Fold</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sig</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also produce a boxplot of normalised counts, which shows the distribution of counts in the whole dataset, and on the subset of peaks with significant differential binding (both upwards and downwards).</p>
<div class="cell" data-hash="../_cache/boxplot_9cdb6840ee34eeafa126c7b5635bf109">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># boxplot of normalised counts</span></span>
<span><span class="fu">dba.plotBox</span><span class="op">(</span><span class="va">brd4_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/boxplot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In our case, this boxplot is interesting, as it illustrates that, overall, BRD4 in E2-treated cells seems to have higher binding affinity than in control cells. This might either be biologically reasonable, or it may be an issue with data quality (recall the variation in FRiP scores we’ve seen earlier and the fact that the PCA separates samples from different treatments along PC1). As we are not experts in this biological system, we refrain from making further comments on this.</p>
<p>Finally, we can visualise the scaled counts in the differentially bound peaks as a heatmap:</p>
<div class="cell" data-hash="../_cache/heatmap_e9442d6237928baf4e268389dda570c2">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># heatmap of DB peaks</span></span>
<span><span class="fu">dba.plotHeatmap</span><span class="op">(</span><span class="va">brd4_dba</span>, contrast <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                correlations <span class="op">=</span> <span class="cn">FALSE</span>, scale <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/heatmap-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="pipeline" class="level2" data-number="6.4"><h2 data-number="6.4" class="anchored" data-anchor-id="pipeline">
<span class="header-section-number">6.4</span> Pipeline</h2>
<div class="cell" data-hash="../_cache/pipeline_header_a83c788ed4449b807b9832a300b3204a">

</div>
<p>Although we’ve broken our analysis down into individual steps, it’s worth nothing that because each function of the <code>DiffBind</code> workflow always returns the <code>dba</code> object with further elements added to it, we can put it all together using <code>|&gt;</code> pipes:</p>
<div class="cell" data-hash="../_cache/unnamed-chunk-3_cef52bae0cff096abad03280c8ffabf7">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># full pipeline - do not run, it will take too long!</span></span>
<span><span class="va">brd4_dba</span> <span class="op">&lt;-</span> <span class="fu">dba</span><span class="op">(</span>sampleSheet <span class="op">=</span> <span class="va">samplesheet</span><span class="op">[</span><span class="va">samplesheet</span><span class="op">$</span><span class="va">Antibody</span> <span class="op">==</span> <span class="st">"BRD4"</span>, <span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.count</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.normalize</span><span class="op">(</span>normalize <span class="op">=</span> <span class="va">DBA_NORM_LIB</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.contrast</span><span class="op">(</span>reorderMeta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="st">"veh"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.analyze</span><span class="op">(</span>method <span class="op">=</span> <span class="va">DBA_DESEQ2</span>,</span>
<span>              bBlacklist <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>              bGreylist <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract diffbound sites</span></span>
<span><span class="va">brd4_diffbound</span> <span class="op">&lt;-</span> <span class="fu">dba.report</span><span class="op">(</span><span class="va">brd4_dba</span>, th <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As we did before, we can also annotate our differentially bound peaks using <code>ChIPseeker</code>:</p>
<div class="cell" data-hash="../_cache/annotate_peaks_eeb98793b212da28d040873258df366f">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># import gene annotation as a transcript database</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">GenomicFeatures</span><span class="fu">::</span><span class="fu">makeTxDbFromGFF</span><span class="op">(</span><span class="st">"resources/GRCh38.109.gtf.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># annotate</span></span>
<span><span class="va">brd4_diffbound</span> <span class="op">&lt;-</span> <span class="fu">annotatePeak</span><span class="op">(</span><span class="va">brd4_diffbound</span>, </span>
<span>                               tssRegion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3e3</span>, <span class="fl">3e3</span><span class="op">)</span>,</span>
<span>                               TxDb <span class="op">=</span> <span class="va">genes</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">as.GRanges</span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="exercises" class="level2" data-number="6.5"><h2 data-number="6.5" class="anchored" data-anchor-id="exercises">
<span class="header-section-number">6.5</span> Exercises</h2>
<div class="callout-exercise callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<div class="cell" data-hash="../_cache/ex_header_d7453343a48d6f696e36781c6e936b76">

</div>
<p>Run the differential binding analysis for the H2Bub1 ChIP. Because counting reads assigned to each peak takes too long to run, we start by loading a pre-processed file.</p>
<p>Starting from this object, build a pipeline to run the rest of the steps in the analysis. We already provide a code skeleton for you to do this, with some “FIXME” for you to correct.</p>
<details><summary>
Code skeleton
</summary><p>Note, this code is provided in the accompanying course materials script. It’s only shown here for reference.</p>
<div class="cell" data-hash="../_cache/ex_h2bub1_697cc6312e7a767182e0e33ce07bb333">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! run diffbind analysis</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>h2bub1_dba <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"preprocessed/r_objects/h2bub1_dba.count.rds"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  FIXME</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! extract results as a GRanges object</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>brd4_diffbound <span class="ot">&lt;-</span> FIXME </span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! correlation heatmap for the samples</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>FIXME</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! PCA plot</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>FIXME</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! MA plot</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>FIXME </span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! boxplot</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>FIXME</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a><span class="co"># !!!FIX!!! heatmap of differentially bound peaks</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>FIXME</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</details><div class="callout-answer callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The full pipeline is shown here (but we don’t run all of it, as it would take too long):</p>
<div class="cell" data-hash="../_cache/ex_answer_part1_e0ac075570087328c359eda3d5d36568">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># full pipeline</span></span>
<span><span class="va">h2bub1_dba</span> <span class="op">&lt;-</span> <span class="fu">dba</span><span class="op">(</span>sampleSheet <span class="op">=</span> <span class="va">samplesheet</span><span class="op">[</span><span class="va">samplesheet</span><span class="op">$</span><span class="va">Antibody</span> <span class="op">==</span> <span class="st">"H2Bub1"</span>, <span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.count</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.normalize</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.contrast</span><span class="op">(</span>reorderMeta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="st">"veh"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.analyze</span><span class="op">(</span>bBlacklist <span class="op">=</span> <span class="cn">FALSE</span>, bGreylist <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the practical, we start from the pre-processed counts:</p>
<div class="cell" data-hash="../_cache/ex_answer_part2_7490aaf1b5f0f5ea8a1d2b7b521319d2">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># start from pre-processed counts</span></span>
<span><span class="va">h2bub1_dba</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"preprocessed/r_objects/h2bub1_dba.count.rds"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.normalize</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.contrast</span><span class="op">(</span>reorderMeta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="st">"veh"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dba.analyze</span><span class="op">(</span>bBlacklist <span class="op">=</span> <span class="cn">FALSE</span>, bGreylist <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run the differential binding analysis</span></span>
<span><span class="va">h2bub1_diffbound</span> <span class="op">&lt;-</span> <span class="fu">dba.report</span><span class="op">(</span><span class="va">h2bub1_dba</span>, th <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then explore our results:</p>
<div class="cell" data-hash="../_cache/ex_answer_part3_c32047564cc9cc0a19dd04871f3571e0">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># summarise</span></span>
<span><span class="va">h2bub1_diffbound</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">summarise</span><span class="op">(</span>up <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">Fold</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>, down <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">Fold</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 1 row and 2 columns
         up      down
  &lt;integer&gt; &lt;integer&gt;
1      3283      1736</code></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># correlation heatmap for the samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">h2bub1_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/ex_answer_part3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># PCA plot</span></span>
<span><span class="fu">dba.plotPCA</span><span class="op">(</span><span class="va">h2bub1_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/ex_answer_part3-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># MA plot</span></span>
<span><span class="fu">dba.plotMA</span><span class="op">(</span><span class="va">h2bub1_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/ex_answer_part3-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># boxplot</span></span>
<span><span class="fu">dba.plotBox</span><span class="op">(</span><span class="va">h2bub1_dba</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/ex_answer_part3-4.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># heatmap of differentially bound peaks</span></span>
<span><span class="fu">dba.plotHeatmap</span><span class="op">(</span><span class="va">h2bub1_dba</span>, contrast <span class="op">=</span> <span class="fl">1</span>, correlations <span class="op">=</span> <span class="cn">FALSE</span>, scale <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="04-diffbind_files/figure-html/ex_answer_part3-5.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In this case there are more differentially bound peaks, and the boxplot of normalised counts between conditions is much closer between groups (although still different, which again might make biological sense).</p>
</div>
</div>
</div>
</div>
</div>
<div class="cell" data-hash="../_cache/unnamed-chunk-4_c0a9dddf20381e7b6b8ceab70fa833c1">

</div>
<!-- 
## Summary

::: {.callout-tip}
#### Key Points

- Last section of the page is a bulleted summary of the key points
::: 
-->


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../materials/03-profile_heatmaps.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Profile heatmaps</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
      <div class="nav-footer-center">Licensed <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a> <img src="https://creativecommons.org/images/deed/cc_icon_black_x2.png" class="img-fluid" width="25"> Bioinformatics Training Facility</div>
  </div>
</footer>


</body></html>